### DISCUSSION TITLE
Force constants

### DISCUSSION THREAD

**Discussion by jk262626** [nakib/elphbolt#146] [Q&A]
Hi Nakib, 
I used a q-grid of 7x7x5 for the phonon calculation in QE (which uses density functional perturbation theory), which is then fed to EPW to get the el-ph coupling. So I used 7 7 5 for `coarse_qmesh` in the input file. 

I used 3rd-order force constants obtained from hiPhive, where atoms are displaced in a 4x4x3 supercell, and 2nd-order force constants from QE with 7x7x5 supercell. Would this be an any issue?

**Comment by nakib** [comment]
Hi Jiyoung,

`coarse_qmesh` is related to Wannierized quantities. This is related to 2nd order force constants only if you have `use_Wannier_ifc2s = .t.`. (Recall from the theory that the real-space representation of the dynamical matrix is nothing but the 2nd order force constants). However, in `elphbolt`, the default is `use_Wannier_ifc2s = .f.`, in which case the 2nd order force constants from `espresso.ifc2` will be used. In that case `coarse_qmesh` is not really relevant since the parser will read the necessary info directly from `espresso.ifc2`. Internally this quantity is called `scell`. See the following code snippet below from `phonon.f90`.
```fortran
type, extends(particle) :: phonon
     !! Data and procedures related to phonons.
     
     character(len = 2) :: prefix = 'ph'
     !! Prefix idenitfying particle type.
     integer(i64) :: scell(3)
     !! q-mesh used in DFPT or, equivalently, supercell used in finite displencement
     !! method for calculating the 2nd order force constants.
     ...
```

And the quantity is set by the private procedure `read_ifc2` bound to the phonon data type.

Regarding the 2nd question: if `hiPhive`'s 3rd order force constants file's format is identical to `thirdorder`'s, then I don't foresee any problems. Again, the relevant information related to the real space is read directly by `elphbolt`. See subroutine `read_ifc3` of data type `phonon`.

I hope this answers all of your questions.

Best,
Nakib

**Reply by jk262626** [reply]
Thank you so much! That's good to know. 
So it doesn't matter even if I use different supercell size for 2nd and 3rd order fc, right?

**Comment by nakib** [comment]
Yes, that's correct.

**Comment by jk262626** [comment]
Nakib, I want to make sure of more things.

1. The `scell` you mentioned would be 7x7x5 in my case. Is that the same as N_ph in the equation below when calculating dynamical matrix?
![image](https://github.com/user-attachments/assets/cf189b1c-6d9b-4dbb-829c-0a1be3e31423)

2. For the 3rd fc, it is used in calculating ph-ph interaction vertices. The i, j, k are the atoms' label in the supercell that I used for 3rd fcs from `hiPhive` or `thirdorder`?
![image](https://github.com/user-attachments/assets/6d13012d-b693-4e52-886c-1a8e72e4ffa0)

3. I've already done long-range correction for polar material when I calculated 2nd fcs in QE. If I use `polar = .true.` in elphbolt as well, does that mean I am doing correction twice and it would be wrong?

4. If I want to recalculate after changing the `polar = .true.` to false, can I still reuse interaction properties by `read_gq2`, `read_gk2`, `read_V`?

**Reply by nakib** [reply]
>     1. The `scell` you mentioned would be 7x7x5 in my case. Is that the same as N_ph in the equation below when calculating dynamical matrix?

I think so. Simply put, there are N_ph number of primitive unit cells in the supercell used for the phonon calculation. You can check that this is the number of R_ph vectors in your 2nd order FC file. Refer to Eq. 32 and the discussion below it in https://journals.aps.org/prb/pdf/10.1103/PhysRevB.76.165108

>     2. For the 3rd fc, it is used in calculating ph-ph interaction vertices. The i, j, k are the atoms' label in the supercell that I used for 3rd fcs from `hiPhive` or `thirdorder`?

I am not familiar with the `hiPhive` code, but this is the case for the 3rd order FC file generated by `thirdorder`. Note that the sum over `i` is restricted to the central primitive unit cell in the supercell.

>     3. I've already done long-range correction for polar material when I calculated 2nd fcs in QE. If I use `polar = .true.` in elphbolt as well, does that mean I am doing correction twice and it would be wrong?

The typical procedure for capturing the polar correction is that the the long range part of the interaction is first **removed** from the 2nd order FCs that are passed to `elphbolt`. The long range corrections (the so called non-analytic term) will be added **after** generating the dynamical matrix at arbitrary **q** vectors. See discussion in Section 5 of https://iopscience.iop.org/article/10.1088/1361-648X/acd831/pdf. I believe there are some subtle differences related to whether you use a displaced supercell method or DFPT to generate the 2nd order FCs. This is discussed along with the reference in the paper shared here.  At the moment `elphbolt` only supports 2nd order FCs generated with DFPT.

I hope this answers all your questions.

Best,
Nakib

**Comment by nakib** [comment]
Let me get back to you on this sometime next week.
